[FEATURE] Créer un script pour rattrapper l'archivage des anciennes organisations avant la suppression automatique des campagnes / participations etc.... (PIX-17863).

xav-car opened the pull request:
## ❄️ Problème

Maintenant que l'archivage des organisations est synchronisé avec la suppression des campagnes / participations / learners / badges ( non certifiable ) . et l'anonymization des assessments / recommended training. il faut rattrapper les anciennes données. 

## 🛷 Proposition

Créer un script qui s'occupe des organizations archivés.

## ☃️ Remarques

## 🧑‍🎄 Pour tester

`node ./scripts/prod/delete-and-anonymise-previous-archive-organization.js --startArchiveDate 2025-12-22`

pix-bot-github commented:
Choisir les applications à déployer :

- [x] [API](https://api-pr14447.review.pix.fr/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr14447/environment) <!-- pix-api-review -->
- [x] [App (.fr)](https://app-pr14447.review.pix.fr) – [App (.org)](https://app-pr14447.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-app-review-pr14447/environment) <!-- pix-app-review -->
- [x] [Orga (.fr)](https://orga-pr14447.review.pix.fr) – [Orga (.org)](https://orga-pr14447.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-orga-review-pr14447/environment) <!-- pix-orga-review -->
- [ ] [Certif (.fr)](https://certif-pr14447.review.pix.fr) – [Certif (.org)](https://certif-pr14447.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-certif-review-pr14447/environment) <!-- pix-certif-review -->
- [ ] [Junior](https://junior-pr14447.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-junior-review-pr14447/environment) <!-- pix-junior-review -->
- [x] [Admin](https://admin-pr14447.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-admin-review-pr14447/environment) <!-- pix-admin-review -->
- [ ] [API MaDDo](https://pix-api-maddo-review-pr14447.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr14447/environment) <!-- pix-api-maddo-review -->
- [ ] [Audit Logger](https://pix-audit-logger-review-pr14447.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr14447/environment) <!-- pix-audit-logger-review -->

> [!IMPORTANT]
> N'oubliez pas de déployer l'API pour pouvoir accéder aux fronts et/ou à l’API MaDDo.

baptiste-h commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+      describe('#participations / assessments / recommendedTrainings / badges', function () {
+        beforeEach(async function () {
+          archivedOrganizationBeforeDate.activeLearner =
+            databaseBuilder.factory.prescription.organizationLearners.buildOrganizationLearner({
```
```Suggestion
            databaseBuilder.factory.buildOrganizationLearner({
```

**question**: On a 2 `databaseBuilder` pour `organizationLearner` ? (les deux sont utilisés à plusieurs endroits en tout cas)

xav-car replied:
* prescription.learner => tend vers le model futur des données dans attributes
* factory.buildLearner => le builder historique avec les données de toutes les colonnes insérables

baptiste-h commented on api/src/prescription/campaign/domain/models/Campaign.js:
```diff
@@ -89,16 +89,22 @@ class Campaign {
-  delete(userId, isAnonymizationWithDeletionEnabled = false) {
-    if (this.deletedAt) {
+  delete(userId, { isAnonymizationWithDeletionEnabled = false, keepPreviousDeletion = false } = {}) {
+    if (this.deletedAt && !keepPreviousDeletion) {
       throw new DeletedCampaignError();
     }
```
**thought**: J'ai l'impression qu'on modifie pas le test de ce cas d'erreur ?

baptiste-h committed:
feat(api): add option to keep previous deletion informations on campaign deletion

Co-authored-by: Xavier Carron <33637571+xav-car@users.noreply.github.com>

baptiste-h committed:
feat(api): keep previous deletion informations on campaign participation deletion

Co-authored-by: Xavier Carron <33637571+xav-car@users.noreply.github.com>

baptiste-h committed:
feat(api): keep previous deletion informations and fix updatedAt on organizationLearner deletion

Co-authored-by: Xavier Carron <33637571+xav-car@users.noreply.github.com>

baptiste-h committed:
This is a combination of 2 commits.

feat(api): add deleted campaign participations to campaign cascade deletion to ensure anonymization

Co-authored-by: Xavier Carron <33637571+xav-car@users.noreply.github.com>

baptiste-h committed:
feat(api): add script to anonymise previously archived organizations

Co-authored-by: Xavier Carron <33637571+xav-car@users.noreply.github.com>

alicegoarnisson commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+      await databaseBuilder.commit();
+    });
+
+    xdescribe('#campaigns', function () {
```
**todo** typo sur le describe

alicegoarnisson commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+              });
+            });
+
+            it('not anonymise and delete participation of archive organization at date', async function () {
```
**todo** ici on veut tester l'organisation archivée à date et pas l'organisation active, attention, car c'est l'organisation active qui est utilisée dans le test

alicegoarnisson commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+              });
+            });
+
+            it('not anonymise and delete participation of active organization', async function () {
```
**todo** inversion du titre avec le test précédent (voir commentaire plus haut)

alicegoarnisson commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+              const assessmentResults = await knex('assessments')
+                .where({ updatedAt: now })
+                .whereNull('campaignParticipationId');
+              expect(assessmentResults).lengthOf(2);
```
**suggestion** on peut tester les ids aussi pour s'assurer qu'on détache les bons assessments

alicegoarnisson commented on api/tests/integration/scripts/prod/delete-and-anonymise-previous-archive-organization_test.js:
```diff
@@ -0,0 +1,697 @@
+          });
+
+          describe('#recommendedTrainings', function () {
+            it('detach its user recommended trainings', async function () {
```
**question** l'organisation archivée à date n'est pas testée ici ?

alicegoarnisson commented on api/src/prescription/campaign-participation/infrastructure/repositories/campaign-participation-repository.js:
```diff
@@ -83,11 +83,14 @@ const get = async function (id) {
 };
 
-const getByCampaignIds = async function (campaignIds) {
+const getByCampaignIds = async function (campaignIds, { withDeletedParticipation = false }) {
```
**todo**  les tests de cette méthode de repo sont à modifier

alicegoarnisson commented on api/src/prescription/campaign/domain/usecases/delete-campaigns.js:
```diff
@@ -16,12 +16,19 @@ const deleteCampaigns = async ({
   eventLoggingJobRepository,
+  client,
+  userRole,
+  keepPreviousDeletion = false,
```
**todo** les tests du usecase sont à modifier

xav-car committed:
tech(api): add withTransaction on delete campaigns

Co-authored-by: Baptiste Hetet <baptiste.hetet@pix.fr>

xav-car commented:
<img width="1895" height="382" alt="image" src="https://github.com/user-attachments/assets/f033b76e-34ed-403b-a25b-7e7d2f900b97" />
Ca me semble OKAY

pix-service commented:
:tada: This PR is included in version 5.292.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix/releases/tag/v5.292.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
